src/backend/cuda/memory.cpp

```diff
@@ -61,29 +57,37 @@ void printMemInfo(const char *msg, const int device) {
 
 template<typename T>
 uptr<T> memAlloc(const size_t &elements) {
-    size_t size                = elements * sizeof(T);
-    MemoryEventPair me = memoryManager().alloc(size, false);
-    cudaStream_t stream        = getActiveStream();
-    if (me.e) me.e.enqueueWait(stream);
-    return uptr<T>(static_cast<T *>(me.ptr), memFree<T>);
+    size_t size = elements * sizeof(T);
+    memory_event_pair pair =
+        memory_event_pair(memoryManager().alloc(elements * sizeof(T), false));
+    event ev            = event(pair.getEvent());
+    Event &e            = getEvent(ev.get());
```

Comment:
You don't have to - you could just write `Event &e = getEvent(pair.getEvent())` here, but I'm creating an `af::event` so we get the RAII benefit/the event is automatically released when `ev` is out of scope.

One alternative would be to add `enqueueWait` etc to the `af::event` API, but that seems like it might make it too heavyweight.

Reply:
The C++ API is an abstraction above the C API. This kind of breaks from that and brings it into the internals of the library. I would like to avoid usage of the public C++ objects internally.

Perhaps we can create a getEvent object that returns an Event by value instead of by reference.

Reply:
Makes sense — didn't know that was a preference. I'll remove uses of `af::event` here.

> returns an Event by value instead of by reference

I'm not sure of how we could do this - the copy ctor of `EventBase` is intentionally deleted because copying an `EventType` seems scary/implementation dependent based on the queue/event semantics of a backend. Since the underlying `af_buffer_info` still has to hold pointer to some `detail::Event`, we can't get it by value here without copying.

Reply:
Perhaps we should rename getEvent to moveEvent and move the event into scope. I think it has better semantic meaning because the get* function return references. Alternatively, the buffer_info function can delete the event on destruction.

Reply:
I don't think that can work because the `af_event` (i.e. the `detail::Event`) in the `pair` here is heap allocated and `EventBase` has a deleted copy ctor, so we can't move it to a local `detail::Event` on the stack.

Either way, I've removed use of `af::event` and `af::buffer_info` from all internal code, so this looks different now.

Reply:
The move constructor should be able to move it from the heap to the stack. Its just copying pointers to the new data structure and disassociates it from the previous object.

The problem we may run into with this approach is if the custom memory manager still references the returned af_buffer_info. We specify that the buffer_info returned by the alloc call cannot be referenced again. This makes sense because we release the af_buffer_info object in the memAlloc call.

Reply:
Agree that the move ctor for `af_event` can do this, but I don't see how it could work for `detail::Event`, since the `af_event` contains a pointer to heap allocated `detail::Event`, and we'd need to move that to a `detail::Event` onto the stack, which we can't do since we have no copy ctor for `detail::Event` (maybe I'm still missing something?)

> We specify that the buffer_info returned by the alloc call cannot be referenced again

Agree; if I understand correctly, you're saying we stipulate that the memory manager can't retain even referential ownership of an `af_buffer_info` after `alloc` returns. This is reasonable/seems required since we can't be sure of the author of a memory manager will free an `af_buffer_info`, so we'd be stuck in a weird spot trying to prevent a leak

Seems more obvious, but we can also specify the same thing for `unlock`: it takes an `af_event`, but it's responsible for releasing it itself.